x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java

```diff
@@ -0,0 +1,365 @@
+                allocationDeciders,
+                routingNodes,
+                state,
+                info,
+                shardSizeInfo,
+                System.nanoTime()
+            );
+            return StreamSupport.stream(state.getRoutingNodes().unassigned().spliterator(), false)
+                .filter(shard -> canAllocate(shard, allocation) == false)
+                .filter(shard -> cannotAllocateDueToStorage(shard, allocation))
+                .mapToLong(this::sizeOf)
+                .sum();
+        }
+
+        public long storagePreventsRemainOrMove() {
+            RoutingNodes routingNodes = new RoutingNodes(state, false);
+            RoutingAllocation allocation = new RoutingAllocation(
+                allocationDeciders,
+                routingNodes,
+                state,
+                info,
+                shardSizeInfo,
+                System.nanoTime()
+            );
+            List<ShardRouting> candidates = state.getRoutingNodes()
+                .shardsWithState(ShardRoutingState.STARTED)
+                .stream()
+                .filter(shard -> allocationDeciders.canRemain(shard, routingNodes.node(shard.currentNodeId()), allocation) == Decision.NO)
+                .filter(shard -> canAllocate(shard, allocation) == false)
+                .collect(Collectors.toList());
+
+            // track these to ensure we do not double account if they both cannot remain and allocated due to storage.
+            Set<ShardRouting> unmovableShards = candidates.stream()
+                .filter(s -> allocatedToTier(s, allocation))
+                .filter(s -> cannotRemainDueToStorage(s, allocation))
+                .collect(Collectors.toSet());
+            long unmovableBytes = unmovableShards.stream()
+                .collect(Collectors.groupingBy(ShardRouting::currentNodeId))
+                .entrySet()
+                .stream()
+                .mapToLong(e -> unmovableSize(e.getKey(), e.getValue()))
+                .sum();
+
+            long unallocatableBytes = candidates.stream()
+                .filter(Predicate.not(unmovableShards::contains))
+                .filter(s1 -> cannotAllocateDueToStorage(s1, allocation))
+                .mapToLong(this::sizeOf)
+                .sum();
+
+            return unallocatableBytes + unmovableBytes;
+        }
+
+        private boolean allocatedToTier(ShardRouting s, RoutingAllocation allocation) {
+            return nodeTierPredicate.test(allocation.routingNodes().node(s.currentNodeId()).node());
+        }
+
+        /**
+         * Check that disk decider is only decider for a node preventing allocation of the shard.
+         * @return true if and only if a node exists in the tier where only disk decider prevents allocation
+         */
+        private boolean cannotAllocateDueToStorage(ShardRouting shard, RoutingAllocation allocation) {
+            assert allocation.debugDecision() == false;
+            allocation.debugDecision(true);
```

Comment:
Can you leave a comment explaining why we need to enable allocation debugging here?

Reply:
:+1:, 6a9c5cb37b021a45c592c283ebb00078eb1990ef